client/driver/exec.go

```diff
@@ -41,12 +46,41 @@ func (d *ExecDriver) Fingerprint(cfg *config.Config, node *structs.Node) (bool,
 }
 
 func (d *ExecDriver) Start(ctx *ExecContext, task *structs.Task) (DriverHandle, error) {
-	// Get the command
+	// Get the command to be ran, or if omitted, download an artifact for
+	// execution. Currently a supplied command takes precedence, and an artifact
+	// is only downloaded if no command is supplied
 	command, ok := task.Config["command"]
 	if !ok || command == "" {
 		return nil, fmt.Errorf("missing command for exec driver")
 	}
 
+	source, ok := task.Config["artifact_source"]
+	if ok && source != "" {
+		// Proceed to download an artifact to be executed.
+		// We use go-getter to support a variety of protocols, but need to change
+		// file permissions of the resulted download to be executable
+
+		// Create a location to download the artifact.
+		taskDir, ok := ctx.AllocDir.TaskDirs[d.DriverContext.taskName]
+		if !ok {
+			return nil, fmt.Errorf("[Err] driver.Exec: Could not find task directory for task: %v", d.DriverContext.taskName)
+		}
+		destDir := filepath.Join(taskDir, allocdir.TaskLocal)
+
+		artifactName := path.Base(source)
+		artifactFile := filepath.Join(destDir, artifactName)
+		if err := getter.GetFile(artifactFile, source); err != nil {
+			return nil, fmt.Errorf("[Err] driver.Exec: Error downloading source for Exec driver: %s", err)
```

Comment:
@catsby Does the getter library test if the artifact to be downloaded exists on the machine already?
For example, if I am running a NodeJs app on an r38xl machine, Nomad should be able to pack around ~16+ Task Groups on a single machine. It would be nice to not download the same file concurrently and if the artifact can be re-used.

Also during Task Group restarts, etc are we going to re-use the binary? This would allow us to start Tasks faster on machines which are already warmed up.